Update IonAuthModel.php

DrAliRagab opened the pull request:
I think you should have function to resend activation email if the first one was not sent successfully.
Depending on that [code ](https://github.com/benedmunds/CodeIgniter-Ion-Auth/issues/1354)I think I made a working function.

The only problem I can't get identity to be sent to email activation, I tried but I failed.
I don't know how to get the identity field.
Actually I didn't tried that hard because I don't need it at activation email, I just want first and last name.

DrAliRagab committed:
Update IonAuthModel.php

I think you should have function to resend activation email if the first one was not sent successfully.
Depending on that [code ](https://github.com/benedmunds/CodeIgniter-Ion-Auth/issues/1354)I think I made a working function.

The only problem I can't get identity to be sent to email activation, I tried but I failed.
I don't know how to get the identity field.
Actually I didn't tried that hard because I don't need it at activation email, I just want first and last name.

benedmunds commented:
Thanks for this PR, this seems like a great addition.  Here are a couple of changes to consider:

1. Move this method to the library, since it's where the other email logic lives and it's not data related
2. Remove deactivation from this method, it should only send the email (if the user is deactivated)
3. Rename to SendActivationEmail, and refactor the register method to call this method instead of sending it's own email directly

To resend emails I would expect the controller logic to be something like `$this->ionAuth->deactivate($user->id)->sendActivationEmail($user->identity)`.

Also, please combine the changes from #1507 into this PR.  I'll be closing that one.

DrAliRagab committed:
Move SendActivationEmail() to Liberaries

Rename function to SendActivationEmail()
Moved it to Liberaries
Make it return Data or boolean according to config.
Now the function expect identity not email.
Added extra messages

DrAliRagab commented:
I think it is better to keep register and send email activation to separate functions.
And we should keep deactivate so we change the activation link every new request.

DrAliRagab committed:
Update IonAuth.php

benedmunds commented on Libraries/IonAuth.php:
```diff
@@ -292,6 +292,86 @@ public function register(string $identity, string $password, string $email, arra
+	 * @return boolean|array return an array of activation details if CI e-mail validation is enabled
+	 * @author Ali Ragab
+	 */
+	public function SendActivationEmail(string $identity): bool|array
```
```suggestion
	public function sendActivationEmail(string $identity): bool|array
```

benedmunds commented on Libraries/IonAuth.php:
```diff
@@ -292,6 +292,86 @@ public function register(string $identity, string $password, string $email, arra
+		if ($user) {
+
+			// deactivate so the user must follow the activation flow
+			$deactivate = $this->deactivate($user->id);
```
Is this (and the next few lines) needed? You're throwing an error above if the user is active.

DrAliRagab replied:
You are right.
Not logic to keep `if ($user) {` as we checked the identity against database and the user must be founded.

But my point regarding `$deactivate = $this->deactivate($user->id);` is to change the activation code every time we request activation code sending. I think it will be more secure to change activation code every request. Am I right?

May be we can regenerate the token and update database, like this code from \Models\IonAuthModel.php
```php
$token                = $this->generateSelectorValidatorCouple(20, 40);
		$this->activationCode = $token->userCode;

		$data = [
			'activation_selector' => $token->selector,
			'activation_code'     => $token->validatorHashed,
			'active'              => 0,
		];
```

But deactivate function do this job perfectly.

benedmunds replied:
Good point.  You could call the model's deactivate method directly to separate those concerns.

DrAliRagab committed:
Update Libraries/IonAuth.php

Co-authored-by: Ben Edmunds <ben.edmunds@gmail.com>

DrAliRagab committed:
Update IonAuth.php

benedmunds commented:
This is looking great. 

One last thing, can you refactor the register() method to use sendActivationEmail(), that way we're reducing the duplicated code there.

DrAliRagab committed:
Update IonAuth.php

DrAliRagab commented:
While working I want to send email to user if his password changed successfully.
So I wrote `sendEmail()` function.
I think it will decrease the duplicate and may be used for further expansion of this Auth system.
My point is that `Register()` trigger defferent events than `sendActivationEmail()`, So better to keep sending email separate.

DrAliRagab commented:
While working I want to send email to user if his password changed successfully.
So I wrote `sendEmail()` function.
I think it will reduce duplicated code here. Also, may be useful for further expansion of this Auth system.

benedmunds commented on Libraries/IonAuth.php:
```diff
@@ -270,16 +270,14 @@ public function register(string $identity, string $password, string $email, arra
-
-				if ($this->email->send() === true)
-				{
+				$email_sent = $this->sendEmail(
```
```suggestion
				$emailSent = $this->sendEmail(
```

benedmunds commented on Libraries/IonAuth.php:
```diff
@@ -292,6 +290,113 @@ public function register(string $identity, string $password, string $email, arra
+
+		} else {
+
+			$email_sent = $this->sendEmail(
```
```suggestion
			$emailSent = $this->sendEmail(
```

benedmunds commented on Libraries/IonAuth.php:
```diff
@@ -292,6 +290,113 @@ public function register(string $identity, string $password, string $email, arra
+				$data
+			);
+
+			if ($email_sent) {
```
```suggestion
			if ($emailSent) {
```

benedmunds commented on Libraries/IonAuth.php:
```diff
@@ -270,16 +270,14 @@ public function register(string $identity, string $password, string $email, arra
+					$data
+				);
+
+				if ($email_sent) {
```
```suggestion
				if ($emailSent) {
```

benedmunds commented on Libraries/IonAuth.php:
```diff
@@ -292,6 +290,113 @@ public function register(string $identity, string $password, string $email, arra
+
+		if ($this->email->send() === true) {
+			return TRUE;
+		}else{
```
```suggestion
		} else {
```

benedmunds commented on Language/en/IonAuth.php:
```diff
@@ -68,6 +68,9 @@
 	'emailActivate_heading'              => 'Activate account for %s',
 	'emailActivate_subheading'           => 'Please click this link to %s.',
 	'emailActivate_link'                 => 'Activate Your Account',
+	'empty_identity'          			 => 'Some thing wrong. Please, try again.',
```
```suggestion
	'empty_identity'          			 => 'Something went wrong. Please try again.',
```

benedmunds reviewed (changes requested):
Just a couple small edits and this should be ready to merge, thanks so much!

DrAliRagab committed:
Update Libraries/IonAuth.php

Co-authored-by: Ben Edmunds <ben.edmunds@gmail.com>

DrAliRagab committed:
Update Libraries/IonAuth.php

Co-authored-by: Ben Edmunds <ben.edmunds@gmail.com>

DrAliRagab committed:
Update Libraries/IonAuth.php

Co-authored-by: Ben Edmunds <ben.edmunds@gmail.com>

DrAliRagab committed:
Update Libraries/IonAuth.php

Co-authored-by: Ben Edmunds <ben.edmunds@gmail.com>

DrAliRagab committed:
Update Libraries/IonAuth.php

Co-authored-by: Ben Edmunds <ben.edmunds@gmail.com>

DrAliRagab committed:
Update Language/en/IonAuth.php

Co-authored-by: Ben Edmunds <ben.edmunds@gmail.com>

benedmunds commented on Libraries/IonAuth.php:
```diff
@@ -292,6 +290,113 @@ public function register(string $identity, string $password, string $email, arra
+		$message = view($template, $data);
+
+		$this->email->clear();
+		$this->email->clear();
```
```suggestion
```

benedmunds reviewed (changes requested):
Found one more issue.  After this I'll merge.

DrAliRagab committed:
Update Libraries/IonAuth.php

Co-authored-by: Ben Edmunds <ben.edmunds@gmail.com>

DrAliRagab commented:
Good review.

benedmunds commented:
Thanks!
